### PR TITLE
Refacto: Policy & keys data structure

### PR DESCRIPTION
**Pull request by ackRow**

### CODE REVIEW

File: src/abe_policy/dimension.rs [hunk 1]
```diff
@@ -195,20 +234,23 @@ impl Dimension {
...
-            Err(Error::OperationNotPermitted(
+        match self {
+            Self::Unordered(attributes) => {
+                if attributes.contains_key(attr_name) {
+                    Err(Error::OperationNotPermitted(
+                        "Attribute already in dimension".to_string(),
+                    ))
+                } else {
+                    attributes.insert(
+                        attr_name.clone(),
```
Comment (tbrezot) [inline src/abe_policy/dimension.rs:245]: Take ownership of the name.

File: src/abe_policy/dimension.rs [hunk 2]
```diff
@@ -223,15 +265,14 @@ impl Dimension {
...
-        if self.order.is_some() {
-            Err(Error::OperationNotPermitted(
-                "Hierarchical dimension are immutable".to_string(),
-            ))
-        } else {
-            self.attributes
+        match self {
+            Self::Unordered(attributes) => attributes
                 .remove(attr_name)
                 .map(|_| ())
```
Comment (tbrezot) [inline src/abe_policy/dimension.rs:271]: Useless line. Juste ignore the returned value by not returning this expression's value.

File: src/abe_policy/dimension.rs [hunk 3]
```diff
@@ -267,24 +314,24 @@ impl Dimension {
...
+        match self {
+            Self::Unordered(attributes) => {
+                if attributes.contains_key(new_name) {
+                    return Err(Error::OperationNotPermitted(
+                        "New attribute name is already used in the same dimension".to_string(),
+                    ));
+                }
+                match attributes.remove(attr_name) {
+                    Some(attr_params) => {
+                        attributes.insert(new_name.to_string(), attr_params);
```
Comment (tbrezot) [inline src/abe_policy/dimension.rs:326]: Take ownership of the new name.

File: src/abe_policy/policy.rs [hunk 4]
```diff
@@ -156,8 +156,7 @@ impl Policy {
         self.dimensions
             .iter()
             .flat_map(|(dim_name, dim)| {
-                dim.attributes
-                    .keys()
+                dim.get_attributes_name()
```
Comment (tbrezot) [inline src/abe_policy/policy.rs:159]: Maybe rename this to `keys()` to follow the `HashMap` conventions.
```suggestion
                dim.keys()
```

File: src/abe_policy/policy.rs [hunk 5]
```diff
@@ -167,8 +166,7 @@ impl Policy {
     /// Fails if there is no such attribute.
     fn get_attribute(&self, attr: &Attribute) -> Result<&AttributeParameters, Error> {
         if let Some(dim) = self.dimensions.get(&attr.dimension) {
-            dim.attributes
-                .get(&attr.name)
+            dim.get_attribute(&attr.name)
```
Comment (tbrezot) [inline src/abe_policy/policy.rs:169]: Maybe rename this to `get` to follow the `HashMap` conventions:
```suggestion
            dim.get(&attr.name)
```

File: src/abe_policy/policy.rs [hunk 6]
```diff
@@ -179,7 +177,7 @@ impl Policy {
     /// The current value is returned first.
     pub fn attribute_values(&self, attribute: &Attribute) -> Result<Vec<u32>, Error> {
         self.get_attribute(attribute)
-            .map(|attr| attr.rotation_values.iter().rev().copied().collect())
+            .map(|attr| attr.all_rotation_values().rev().collect())
```
Comment (tbrezot) [inline src/abe_policy/policy.rs:180]:
```suggestion
            .map(|attr| attr.values().rev().collect())
```

File: src/abe_policy/policy.rs [hunk 7]
```diff
@@ -363,8 +360,7 @@ fn generate_current_attribute_partitions(
         if !current_attr_value_per_dim.contains_key(dim) {
             // gather all the latest value for that dim
             let values = dim_properties
-                .attributes
-                .values()
+                .attributes_properties()
```
Comment (tbrezot) [inline src/abe_policy/policy.rs:363]: You want to get the attributes?
```suggestion
                .attributes()
```
You can also use `values()` to follow the `HashMap` conventions (a dimension is a map from names to attributes):
```rust
                .values()
```

File: src/abe_policy/dimension.rs [hunk 8]
```diff
@@ -123,33 +138,49 @@ impl Dimension {
...
+        }
+    }
+
+    /// Returns an iterator over the attributes name.
+    /// If the dimension is ordered, the names are returned in this order,
+    /// otherwise they are returned in arbitrary order.
+    pub fn get_attributes_name(&self) -> Box<dyn '_ + Iterator<Item = &AttributeName>> {
+        match self {
+            Self::Unordered(attributes) => Box::new(attributes.keys()),
+            Self::Ordered(attributes) => Box::new(attributes.keys()),
```
Comment (ThibsG) [inline src/abe_policy/dimension.rs:176]: note that if `attributes` is the same type on each arm you can write:
```suggestion
            Self::Unordered(attributes) | Self::Ordered(attributes) => Box::new(attributes.keys()),
```

File: src/abe_policy/dimension.rs (hunk 2 above)
Comment (ackRow) [reply src/abe_policy/dimension.rs:223]: So we will not raise an exception if the Attribute wasn't removed?

File: src/abe_policy/policy.rs (hunk 4 above)
Comment (ackRow) [reply src/abe_policy/policy.rs:146]: `Dimension` is an abstraction so I am not sure it's a good idea to use the same names as the underlying structs (which could change)

**Review by tbrezot — APPROVED** [review]

**Pull request merged**